### ISSUE TITLE
[API Proposal]: Add `Stream` wrappers for memory and text-based types

### ISSUE THREAD

**Issue by jozkee** [dotnet/runtime#82801]
### Background and motivation

This is meant to bundle https://github.com/dotnet/runtime/issues/46663, https://github.com/dotnet/runtime/issues/27156, and https://github.com/dotnet/runtime/issues/22838.

There's the usual need of wrapping memory and text-based types in a Stream and given that there are multiple implementations spread across different libraries, it would be ideal to have a standardized API for it directly on .NET.

A few examples of the current options available:
[CommunityToolkit.HighPerformance](https://github.com/CommunityToolkit/dotnet) provides `AsStream()` extension methods for the following:
`Memory<byte>`
`ReadOnlyMemory<byte>`
`IMemoryOwner<byte>`
`IBufferWriter<byte>`

[Nerdbank.Streams](https://github.com/AArnott/Nerdbank.Streams/blob/main/doc/AsStream.md) also provides `AsStream()` extensions for the following:
`IDuplexPipe`
`PipeReader/PipeWriter`
`WebSocket`
`ReadonlySequence<byte>`
`IBufferWriter<byte>`

While I'm not certain of providing support for all the types listed above, I think a good starting point would be the following based on the evidence shown in the issues I bundled.

### API Proposal
Originally proposed by @bartonjs on https://github.com/dotnet/runtime/issues/82801#issuecomment-1449200862.
Static factory methods on `System.IO.Stream` for CoreLib types, with a C#14 extension type in System.Memory.dll for `ReadOnlySequence<byte>`.

```cs
// In System.Private.CoreLib (System.Runtime.dll)
namespace System.IO
{
    public partial class Stream
    {
        public static Stream FromText(string text, Encoding? encoding = null);
        public static Stream FromText(ReadOnlyMemory<char> text, Encoding? encoding = null);

        public static Stream FromReadOnlyData(ReadOnlyMemory<byte> data);
        public static Stream FromWritableData(Memory<byte> data);
    }
}

// In System.Memory.dll (C#14 extension type)
namespace System.IO
{
    public static class ReadOnlySequenceStreamExtensions
    {
        extension(Stream)
        {
            public static Stream FromReadOnlyData(ReadOnlySequence<byte> sequence);
        }
    }
}
```
**Pros:**
- Unified developer experience - all methods appear as `Stream.From*()` via IntelliSense, regardless of underlying assembly
- Solves layering cleanly - C#14 extensions let System.Memory.dll contribute statics to `Stream` without modifying CoreLib
- Maximum discoverability (FDG: "frameworks must be usable without the need for documentation")

**Cons:**
- Not all .NET languages support C#14 extension syntax; callers in those languages would use `ReadOnlySequenceStreamExtensions.FromReadOnlyData(...)` instead of the projected `Stream.FromReadOnlyData(...)`
-  Statics on `Stream` in CoreLib are only available to .NET 10+ consumers; multi-targeting libraries (netstandard2.0) cannot use them

### API Usage

Example 1:
```cs
static HttpClient client = new HttpClient();
static void SendString(string str)
{
    var request = new HttpRequestMessage(HttpMethod.Post, "contoso.com");
    request.Content = new StreamContent(Stream.FromText(str));
    client.Send(request);
}
```

Example 2:
```cs
static unsafe void DoStreamOverSpan(ReadOnlySpan<byte> span)
{
    fixed (byte* ptr = &MemoryMarshal.GetReference(span))
    {
        using (MemoryManager<byte> manager = new PointerMemoryManager<byte>(ptr, span.Length))
        {
            using Stream s = Stream.FromReadOnlyData(manager.Memory);
            // Do something with it...
        }
    }
}
```

Example 3:

```cs
// Pipeline protocol deserialization with System.IO.Pipelines
ReadResult result = await pipeReader.ReadAsync();
ReadOnlySequence<byte> buffer = result.Buffer;

// Stream-only deserializer (e.g., XmlSerializer, legacy APIs)
using var stream = Stream.FromReadOnlyData(buffer);
var config = (AppConfig)new XmlSerializer(typeof(AppConfig)).Deserialize(stream);

pipeReader.AdvanceTo(buffer.End);
```

### Alternative Design 1

Static factory methods on `System.IO.Stream` for CoreLib types, with a classic extension method on `ReadOnlySequence<byte>` in System.Memory.dll.

```cs
// In System.Private.CoreLib (System.Runtime.dll)
namespace System.IO
{
    public partial class Stream
    {
        public static Stream FromText(string text, Encoding? encoding = null);
        public static Stream FromText(ReadOnlyMemory<char> text, Encoding? encoding = null);

        public static Stream FromReadOnlyData(ReadOnlyMemory<byte> data);
        public static Stream FromWritableData(Memory<byte> data);
    }
}

// In System.Memory.dll (classic extension method)
namespace System.Buffers
{
    public static class ReadOnlySequenceExtensions
    {
        public static Stream AsStream(this ReadOnlySequence<byte> sequence);
    }
}
```

**Pros:**
- Uses only well-established API patterns
    - The use of C#14 static extensions in BCL public API is a new pattern still being evaluated by the review board.
- `AsStream()` follows existing precedent (`IEnumerable<T>.ToList()`, `HttpContent.ReadFromJsonAsync()` - FDG 5.6)
- Works in all .NET languages (no extension member syntax dependency)

**Cons:**
- Inconsistent API pattern: `Stream.FromReadOnlyData(memory)` vs `sequence.AsStream()` for the same conceptual operation
- `AsStream()` is less discoverable — requires knowing to look on `ReadOnlySequence<byte>` instead of `Stream`
- In case we want to go all-in with `AsStream()` for all types, we cannot extend `string` [FDG 5.6]: Besides injecting an unrelated concept into a fundamental type, since `string` is one of the most common types, the pollution in IntelliSense would be massive.

### Alternative Design 2

Move `ReadOnlySequence<T>` to System.Private.CoreLib to enable adding the stream wrapper as a static method directly on `Stream`, consistent with the other CoreLib types.

From https://github.com/dotnet/runtime/issues/82801#issuecomment-1456497703 and https://github.com/ViveliDuCh/runtime/pull/1#discussion_r2717398055

```cs
// In System.Private.CoreLib (System.Runtime.dll)
// Note: This requires moving ReadOnlySequence<T> from System.Memory.dll to System.Private.CoreLib
namespace System.IO
{
    public partial class Stream
    {
        public static Stream FromText(string text, Encoding? encoding = null);
        public static Stream FromText(ReadOnlyMemory<char> text, Encoding? encoding = null);

        public static Stream FromReadOnlyData(ReadOnlyMemory<byte> data);
        public static Stream FromReadOnlyData(ReadOnlySequence<byte> sequence);
        public static Stream FromWritableData(Memory<byte> data);
    }
}
```

**Pros:**
- Fully consistent — all methods are real statics on `Stream`, single namespace and entry point
- Maximum discoverability via IntelliSense (FDG: "frameworks must be usable without the need for documentation")

**Cons:**
- Requires moving `ReadOnlySequence<T>` to CoreLib - too much change for a single API scenario
- Breaks `ReadOnlySequence<T>` availability on older targets (currently available via NuGet on netstandard2.0)

### Alternative Design 3

Place all factory methods in a `StreamFactory` class in System.Memory.dll (which depends on CoreLib where `System.IO.Stream` lives).

As suggested by @adamsitnik in [this comment](https://github.com/ViveliDuCh/runtime/pull/1#discussion_r2708780320).

```cs
// In System.Memory.dll
namespace System.IO
{
    public static class StreamFactory
    {
        public static Stream FromText(string text, Encoding? encoding = null);
        public static Stream FromText(ReadOnlyMemory<char> text, Encoding? encoding = null);

        public static Stream FromReadOnlyData(ReadOnlyMemory<byte> data);
        public static Stream FromReadOnlyData(ReadOnlySequence<byte> sequence);

        public static Stream FromWritableData(Memory<byte> data);
    }
}
```

**Pros:**
- No layering issues — all methods in one class in System.Memory.dll, which can reference both CoreLib and System.Memory types
- Broader reach potential if System.Memory.dll ships as a NuGet for older frameworks

**Cons:**
- Lower discoverability — developers look for stream creation on `Stream`, not on `StreamFactory` (FDG: "Factories sacrifice discoverability, usability, and consistency for implementation flexibility")
- No IntelliSense guidance when typing `Stream.` - users must know to look for a separate class
- Generally less preferred when constructors/static methods are viable

### Locations

The following dotnet/runtime production code could consume this API:

| Library | File | Usage |
|---------|------|--------------|
| System.Net.Http | [ReadOnlyMemoryContent.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http/src/System/Net/Http/ReadOnlyMemoryContent.cs#L35) | ✅ Yes |
| System.Private.Xml | [XmlPreloadedResolver.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.Xml/src/System/Xml/Resolvers/XmlPreloadedResolver.cs#L106) | ✅ Yes |
| System.Private.DataContractSerialization | [JsonXmlDataContract.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/JsonXmlDataContract.cs#L33) |✅ Yes |
| System.Memory.Data | [BinaryData.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Memory.Data/src/System/BinaryData.cs#L425) | ❌ No |
| System.Configuration.ConfigurationManager | [ImplicitMachineConfigHost.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs#L68) | ❌ No |
| System.Resources.Extensions | [DeserializingResourceReader.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Resources.Extensions/src/System/Resources/Extensions/DeserializingResourceReader.cs#L218) | ❌ No |

**Current Limitation:** Libraries multi-targeting `netstandard2.0`/`net462` cannot use the CoreLib-based API for older targets.

### Notes

- **ReadOnlySequence\<byte\> wrapper need has evolved:** Since [#27156](https://github.com/dotnet/runtime/issues/27156) was filed (2018), most major serializers (`System.Text.Json`, `MessagePack`, `protobuf-net`) now accept `ReadOnlySequence<byte>` directly. The remaining gap is Stream-only APIs (`XmlSerializer`, legacy deserializers) and frameworks like HybridCache ([#100434](https://github.com/dotnet/runtime/issues/100434)) where `Stream` remains the most common serializer fallback.
    - `PipeReader.AsStream()` does not cover this gap — `PipeReader.Create(sequence).AsStream()` produces a non-seekable stream, while consumers like `XmlSerializer` and `DataContractSerializer` use `Length`/`Seek` for buffer sizing and format detection. A purpose-built wrapper provides seekable, zero-copy access matching in-memory stream semantics.

- **Partial coverage of [#100434](https://github.com/dotnet/runtime/issues/100434):** This proposal addresses the `ReadOnlySequence<byte>` → `Stream` portion of that issue. The `IBufferWriter<byte>` → `Stream` wrapper is not included here and would be a separate proposal.

### Open questions

1. **Naming consistency:** Does the `Stream.From*` naming present a coherent mental model for stream creation across text, memory, and sequences?
2. Should the `ReadOnlySequence<byte>` overload use a C#14 static extension on `Stream` or a classic extension method (e.g., `AsStream()`) in System.Memory.dll? See [PR discussion](https://github.com/ViveliDuCh/runtime/pull/1#discussion_r2722518577), [preliminary review](https://github.com/dotnet/runtime/issues/82801#issuecomment-3916837306), and [subsequent discussion](https://github.com/dotnet/runtime/issues/82801#issuecomment-3918100849).
3. If C#14 static extension approach is used, the fallback invocation for non-C#14 languages becomes `ReadOnlySequenceStreamExtensions.FromReadOnlyData(...)`, does the extension class name need to follow a naming convention that preserves discoverability alongside `Stream.FromReadOnlyData(...)`?
4. If classic extension method is preferred, is the naming asymmetry (`Stream.FromReadOnlyData(ROM<byte>)` vs `sequence.AsStream()`) acceptable?
5. [[Based on preliminary api review] ](https://github.com/dotnet/runtime/issues/82801#issuecomment-3916837306) **Parameter naming:** Should the `encoding` parameter in `FromText` be renamed to `streamEncoding` to disambiguate its purpose? i.e.  `Stream.FromText(myString, streamEncoding: Encoding.Latin1)`      - The parameter specifies the encoding used for the bytes in the *resulting stream*, not the encoding of the *source text*.
6. Should `FromText` expose an optional `bufferSize` parameter (following [`StreamReader`'s constructor pattern](https://github.com/dotnet/dotnet/blob/b0f34d51fccc69fd334253924abd8d6853fad7aa/src/runtime/src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs#L138)), or keep the encoding buffer size as a fixed internal detail?

**Comment by Jozkee** [comment]
cc @AArnott @Sergio0694 @stephentoub.

**Comment by bartonjs** [comment]
I don't think that making a stream from a string (or byte buffer) is common enough to warrant an extension method.  I'd rather see it as static non-extension methods on Stream, such as

```C#
public partial class Stream
{
    public static Stream FromText(string text, Encoding encoding);
    public static Stream FromText(ReadOnlyMemory<char> text, Encoding encoding);

    public static Stream FromData(byte[] data);
    public static Stream FromData(ReadOnlyMemory<byte> data);
    public static Stream FromData(ReadOnlySequence<byte> data);
}
```

**Comment by Jozkee** [comment]
I don't know how common is too common but the [top StackOverflow result](https://stackoverflow.com/questions/1879395/how-do-i-generate-a-stream-from-a-string) has 1.2k upvotes in the top answer and is quite suboptimal.

```cs
public static Stream GenerateStreamFromString(string s)
{
    var stream = new MemoryStream();
    var writer = new StreamWriter(stream);
    writer.Write(s);
    writer.Flush();
    stream.Position = 0;
    return stream;
}
```

**Comment by AArnott** [comment]
When it comes to converting text to a Stream, I don't know that your anti-sample is all that bad. You'll have to encode the text either way. I suppose you could encode it in a streaming fashion instead of all up front, but that would add a great deal of complexity, and for a string (which _shouldn't_ be too large in the common case) the complexity may make it slower.

**Comment by Jozkee** [comment]
@AArnott that's fair. It's possible that encoding on the go is not more optimal in all scenarios, especially on small strings as you say.

**Comment by skyoxZ** [comment]
> I don't think that making a stream from a string (or byte buffer) is common enough to warrant an extension method. I'd rather see it as static non-extension methods on Stream

Agree. And the extension methods are under such a common namespace `System.IO`, which means they are always there.

**Comment by dersia** [comment]
Since most common cases will be with utf8 strings and since we ha e utf8 literals I think another `FromUtf8String(string/rom<char> text)` would be very useful.

**Comment by stephentoub** [comment]
> Since most common cases will be with utf8 strings and since we ha e utf8 literals I think another FromUtf8String(string/rom<char> text) would be very useful.

How does that differ from MemoryStream of you already have a byte[], or the proposes FromMemory of you have a ROM?

**Comment by dersia** [comment]
Maybe I wasn't clear enough, I am not talking about the FromMemory methods, you are right, that it is easy to just use those after getting the bytes for a utf8 string. I am referring to the two FromText methods and I meant having a FromUtf8Text method overloads that don't require an encoding.

**Comment by stephentoub** [comment]
> meant having a FromUtf8Text method overloads that don't require an encoding.

You mean so you can write:
```C#
Stream s = Stream.FromUtf8Text(str);
```
instead of:
```C#
Stream s = Stream.FromText(text, Encoding.UTF8);
```
?

**Comment by dersia** [comment]
Yeah, I understand how ridiculous this sound, since it is a very simple Encoding.UTF8 what I am asking to leave out, and this is my thinking when I read about the BinaryData Proposal, but in my experience this has helped many coming from other languages.

Maybe it should be just an overload instead of it's own method and also be called FromText and just assume if encoding isn't specified it is utf8.

**Comment by svick** [comment]
I'm not quite certain about the naming. Especially since the difference between `new MemoryStream(byteArray)` and `Stream.FromData(byteArray)` is not at all clear from the latter's name to me.

What about `Stream.ReadData` and `Stream.ReadText`? Though that name is quite similar `File.ReadAllText`, which does something fairly different.

**Comment by dersia** [comment]
> I'm not quite certain about the naming. Especially since the difference between `new MemoryStream(byteArray)` and `Stream.FromData(byteArray)` is not at all clear from the latter's name to me.
> 
> What about `Stream.ReadData` and `Stream.ReadText`? Though that name is quite similar `File.ReadAllText`, which does something fairly different.

Maybe FromBytes? I think Read has a different meaning, to me it would imply that I am reading data or a text from the stream, but the From is used in the bcl for creating new instances, so it makes sense to stay with that pattern.

I think it would make sense to have an extension coming from STJ on stream that creates a stream from an object and that would than be called FromData<T>, although FromObject<T> sounds more like it...

**Comment by koszeggy** [comment]
I would suggest making the `encoding` parameter nullable, or adding  `FromText` overloads without the encoding parameters.

If the encoding is not specified, the methods would return the most performant and trivial stream _wrapper_ (as the title suggest) instead of a more complex _transcoder_:

```cs
public partial class Stream
{
    public static Stream FromText(string text); // a StreamReader for the result can use Encoding.Unicode
    public static Stream FromText(ReadOnlyMemory<char> text); // Encoding.Unicode
    public static Stream FromText(byte[] text); // Encoding.Utf8
    public static Stream FromText(ReadOnlyMemory<byte> text); // Encoding.Utf8

    [...] // and the originally proposed overloads with the encoding parameter
}
```

Feel free to disagree - I also see that in the SO topic (mentioned by @Jozkee) my [answer](https://stackoverflow.com/a/57100948/5114784) is far less popular than the accepted one, even though it's the only one, which is an actual wrapper. Of course, transcoders also can be implemented without reading the whole source in advance but even then they obviously have some overhead both in performance and allocation.

**Comment by skyoxZ** [comment]
I think there's no problem of extension method `AsStream` for bytes sequence such as `byte[]`. It's a little more convenient and clear: what happens if I alter the array after `Stream.FromData(byte[] data)`?

**Comment by stephentoub** [comment]
> public static Stream FromData(ReadOnlySequence<byte> data);

@Jozkee, note this proposed API isn't possible unless `ReadOnlySequence<T>` is moved into Corelib.

**Comment by davidfowl** [comment]
> @Jozkee, note this proposed API isn't possible unless ReadOnlySequence<T> is moved into Corelib.

😍

**Comment by davidfowl** [comment]
> We could rule-out ReadOnlySequence from the proposal and provide a separate solution.

Is there a reason not to do this?

**Comment by dersia** [comment]
> > I would suggest making the encoding parameter nullable, or adding FromText overloads without the encoding parameters.
> 
> @koszeggy our current stance is to force users to reason about encoding, see [#46663 (comment)](https://github.com/dotnet/runtime/issues/46663#issuecomment-1449197413). I will also add that it is easy to relax it in the future (by adding an overload or making the param nullable as you suggest), but is hard to go the opposite way.

I am not even sure what a null Encoding would mean. if it is in case of null defaulted to Unicode, shouldn't it instead than just be defaulted to Unicode? I understand, that this is out of scope for now, I am just wondering, if it would make sense at all to have it nullable and if what the semantic of a nullable encoding would mean here.

as for `FromText(byte[] text)` defaulting to utf8, I don't think it is neededn, becuase in this case we can just use `FromData("this is text"u8);` 

FWIW:
I just wanted to raise the concern, that Streams are mostly used in a networking environment and very often for the web and since UTF8 is mostly used on the web, it makes sense to default to UTF8 encoding. I understand, that this isn't the most performant encoding and there are cases where performance is most important, I think that most use cases of this api won't be optimizing towards performance.
This is why I suggested to default to UTF8 just for Developer convenience. If one wants to optimize for performance, than there is nothing stopping them from using `FromText("this is some text", Encoding.Unicode)`. The BCL is used from every dev so optimizing for convenience over optimizing for performance makes more sense in such a High-Level API.

**Comment by Jozkee** [comment]
> > We could rule-out ReadOnlySequence from the proposal and provide a separate solution.
>
> Is there a reason not to do this?

Only consistency I think. I interpreted @stephentoub's comment on https://github.com/dotnet/runtime/issues/27156#issuecomment-431352364 as an ask to bring a consistent solution between `[ReadOnly]Memory<T>` and `ReadOnlySequence<T>`.

**Comment by bartonjs** [comment]
> >    note this proposed API isn't possible unless ReadOnlySequence is moved into Corelib
>
> ah, I haven't thought of that. This wouldn't be a problem for extension methods.

It's not extension methods vs not, it's "on Stream" or "on a sibling type".  Extension methods would be forced to be on a new type; but it could just as easily be the same proposal, but as non-extension methods on a new type, which lives in a higher ring.

```C#
public class StreamFactory
{
    public static Stream FromText(string text, Encoding encoding);
    public static Stream FromText(ReadOnlyMemory<char> text, Encoding encoding);

    public static Stream FromData(byte[] data);
    public static Stream FromData(ReadOnlyMemory<byte> data);
    public static Stream FromData(ReadOnlySequence<byte> data);
}
```

(The "From" name seems wrong there, but it could be massaged...)

**Comment by Jozkee** [comment]
> It's not extension methods vs not, it's "on Stream" or "on a sibling type"

@bartonjs yes, I got that, I just thought extensions still had a chance over factories. 
Anyway, your proposal seems nicer, and upvotes had already shown that there's more interest on a factory approach.

I will update the proposal shortly.

**Comment by svick** [comment]
@Jozkee 

> @svick the returned stream from Strem.FromData(byte[]) should be capable of writing, so conveying "Read" on the name is not possible.

Then what is the difference between `new MemoryStream(byteArray)` and  `Stream.FromData(byteArray)`? Or does that overload only exist for consistency?

Also, what makes `byte[]` special? Why not also include overloads for `Memory<byte/char>`, if writing is supported for `byte[]`?

**Comment by Jozkee** [comment]
> what makes byte[] special?

@svick that's my bad, I didn't notice @bartonjs proposal included `FromData(byte[] data)`, I think that should be `FromData(Memory<byte> data)` and the implicit operators will allow passing byte[].

I'm not sure how useful `FromText(Memory<char>)` could be but will include it as well for consistency.

@bartonjs, objections?

**Comment by bartonjs** [comment]
There's always the question of when we stop, but we've been leaving `T[]` overloads when we take Memory/ReadOnlyMemory/Span/ReadOnlySpan in "common" "friendly" APIs on the basis of both language interop concerns and not making CS-1-equivalent programmers need to start off by understanding what Span/Memory are right off the bat.

> I'm not sure how useful FromText(Memory<char>) could be but will include it as well for consistency.

Guidelines say not to have both a ReadOnlyMemory and Memory overload in the same method group (unless it's an extension method).

**Comment by Jozkee** [comment]
> Guidelines say not to have both a ReadOnlyMemory and Memory overload in the same method group (unless it's an extension method).

But in this case, Memory and ReadOnlyMemory convey different things. that one is writable and readable and the other is readable only. Can't this be an exception to the rule?

**Comment by bartonjs** [comment]
Well, when two overloads have different semantics the guidelines suggest they shouldn't be overloads, but different names (different method groups).

I had thought this was only about read-only streams.  If writing is desired, then perhaps it becomes something more like

```C#
public class StreamFactory
{
    public static Stream CreateReadOnly(string text, Encoding encoding);
    public static Stream CreateReadOnly(ReadOnlyMemory<char> text, Encoding encoding);
    public static Stream CreateReadOnly(byte[] data);
    public static Stream CreateReadOnly(ReadOnlyMemory<byte> data);
    public static Stream CreateReadOnly(ReadOnlySequence<byte> data);

    public static Stream CreateWriteOnly(Memory<char> buffer, Encoding encoding);
    public static Stream CreateWriteOnly(byte[] buffer);
    public static Stream CreateWriteOnly(Memory<byte> buffer);

    // Would these need knobs to control the read-pos and write-pos differently?
    public static Stream CreateReadWrite(Memory<char> buffer, Encoding encoding);
    public static Stream CreateReadWrite(byte[] buffer);
    public static Stream CreateReadWrite(Memory<byte> buffer);
}
```

**Comment by Jozkee** [comment]
> // Would these need knobs to control the read-pos and write-pos differently?

I don't think so, all streams just rely in one Position. If you start Reading 10 bytes and then you Write 10 bytes, you end up on Position 20. The same should happen here.

**Comment by bartonjs** [comment]
There were two main approaches discussed in review.

For the first, having `FromText` and `FromData` methods on `Stream`.  For ReadOnlySequence this would either be to move the type down, or to defer it to the proposed static-extensions C# feature (to allow a higher layer to project a `static Stream.FromData(ReadOnlySequence<byte>)` down).

The second was to see if MemoryStream can be changed (without significant penalty) to work off of Memory/ReadOnlyMemory. If it can, then we'd move `FromText` to something off of System.Text.Encoding (and we'd do something like `AsStream()` for ReadOnlySequence).

The second seemed to have a bit more favor, but requires further investigation, so classifying this as needs-work.

```C#
// It will be in a new assembly with a dependency on System.Memory.dll and System.Runtime.dll
namespace System.IO
{
    public partial class StreamFactory
    {
        public static Stream FromText(string text, Encoding encoding);
        public static Stream FromText(Memory<char> text, Encoding encoding);
        public static Stream FromText(ReadOnlyMemory<char> text, Encoding encoding);

        public static Stream FromData(Memory<byte> data);
        public static Stream FromData(ReadOnlyMemory<byte> data);
        public static Stream FromData(ReadOnlySequence<byte> data);
    }
}
```

**Comment by jeffhandley** [comment]
We're moving this out to Future and it will be considered again for .NET 9.

**Comment by bartonjs** [comment]
API Review (preliminary)

There was pretty strong agreement against static extension methods.  The most favorable approach was to move ROSequence into Corelib.

People were surprised at UTF-8 being the default encoding for FromText, but it might be correct (see also WriteAllText).

People thought maybe `streamEncoding` instead of `encoding` for the parameter name in FromText.

**Comment by tannergooding** [comment]
Thanks Vivi. Just a few clarifications below, which will be relevant when this comes back to API review.

> "Unspeakable name" / no static fallback: Static extension methods are emitted by the compiler under a name that isn't valid C# syntax. Unlike classic extension methods (which you can always call as ExtensionClass.Method(...)), static extensions can only be invoked through the projected syntax (Stream.FromReadOnlyData(...)). There is no alternative calling pattern.

This is incorrect. Static extension methods are invocable as simply `ExtensionClass.Method`, the "unspeakable name" is an internal implementation detail only relevant to metadata readers (like Roslyn) to know what type to expose such APIs on. 

In other words, this is valid:
```csharp
Console.WriteTab();          // extension invocation
ConsoleExtension.WriteTab(); // explicit invocation

public static class ConsoleExtension
{
    extension(Console)
    {
        public static void WriteTab() => Console.Write('\t');
    }
}
```

> Not a user-repeatable pattern: The guidelines say don't extend types you don't own. If an external library came up with a type like ReadOnlySequence and added a static extension on Stream, we'd tell them they shouldn't.

This sounds like a reason to figure out some instance method invocation pattern as well, like the alternatively proposed `ToStream(...)`

> Name stomping is irreversible: Since static extensions can only be called via the extended type's syntax, if a real member with the same signature is later added to the type, the extension becomes literally uncallable. (No fallback syntax to reach the shadowed extension)

This is incorrect, as per the response to 1

> Runtime binding degradation: The compiler resolves static extensions fine, but runtime binding does not.

Not sure I understand this one or how it relates. The experience is the same as all extension APIs, such as all of LINQ
